File: computer-graphics/bresenham-line.py
from pprint import pprint

class Bresenham:
  def __init__(self, cord):
    self.x1, self.y1, self.x2, self.y2 = cord
    
    self.output = [[self.x1, self.y1]]
    
  def run(self):
    delX = self.x2 - self.x1
    delY = self.y2 - self.y1

    slope = delY / delX
    
    currentP = 2 * delY - delX
    
    if slope >= 1:
      currentP = 2 * delX - delY
    
    currentX = self.x1
    currentY = self.y1
    
  
    while ((slope < 1 and currentX < self.x2) or
      (slope >= 1 and currentY < self.y2)):
      
      if slope < 1 and currentP < 0:
        currentP += 2 * delY
        currentX += 1
      elif slope < 1 and currentP >= 0:
        currentP += 2 * delY - 2 * delX
        currentY += 1
        currentX += 1
      elif slope >= 1 and currentP < 0:
        currentP += 2 * delX
        currentY += 1
      elif slope >= 1 and currentP >= 0:
        currentX += 1
        currentY += 1
        currentP += 2 * delX - 2 * delY

      self.output.append([currentX, currentY])

    return self.output
  
    # if slope < 1:
    #   while(currentX < self.x2):
        
    #     if currentP < 0:
    #       currentP = currentP + 2 * delY
    #     else:
    #       currentP = currentP + 2 * delY - 2 * delX
    #       currentY += 1
          
    #     currentX += 1

    #     print(currentX, currentY)

    #     self.output.append([currentX, currentY])
        
    # else: 
    #    while(currentY < self.y2):
        
    #     if currentP < 0:
    #       currentP = currentP + 2 * delX
    #     else:
    #       currentP = currentP + 2 * delX - 2 * delY
    #       currentX += 1
          
    #     currentY += 1
    #     self.output.append([currentX, currentY])
        
    # return self.output
  
inputVal = list(map(int, input("Enter cordinates ").split()))

pprint(Bresenham(inputVal).run())

File: computer-graphics/bresenham-line.js
class Bresenham {
  constructor(cord1, cord2) {
    this.cord1 = cord1;
    this.cord2 = cord2;
    this.output = [this.cord1];
  }

  run() {
    console.log(
      `(${this.cord1[0]}, ${this.cord1[1]}), (${this.cord2[0]}, ${this.cord2[1]})`
    );
    
    const delX = this.cord2[0] - this.cord1[0],
      delY = this.cord2[1] - this.cord1[1];

    const slope = delY / delX;

    console.log(`delX = ${this.cord2[0]} - ${this.cord1[0]} = ${delX}`);
    console.log(`delY = ${this.cord2[1]} - ${this.cord1[1]} = ${delY}`);

    
    let currentP = slope < 1 ? 2 * delY - delX : 2 * delX - delY;
    let currentX = this.cord1[0],
    currentY = this.cord1[1];

    console.log(`slope == ${slope}`);
    
    
    if (slope < 1) console.log(`p = 2 * ${delY} - ${delX} = `, currentP);
    else console.log(`p = 2 * ${delX} - ${delY} = `, currentP);

    while (
      (delX >= delY && currentX < this.cord2[0]) ||
      (delX < delY && currentY < this.cord2[1])
    ) {
      let operatorStr = "";

      if (slope < 1 && currentP < 0) {
        operatorStr = `${currentP} + 2 * ${delY}`;
        currentP += 2 * delY;
        currentX++;
      } else if (slope < 1 && currentP >= 0) {
        operatorStr = `${currentP} + 2 * ${delY} - 2 * ${delX}`;
        currentP += 2 * delY - 2 * delX;
        currentX++;
        currentY++;
      } else if (slope >= 1 && currentP < 0) {
        operatorStr = `${currentP} + 2 * ${delX}`;
        currentP += 2 * delX;
        currentY++;
      } else if (slope >= 1 && currentP >= 0) {
        operatorStr = `${currentP} + 2 * ${delX} - 2 * ${delY}`;
        currentX++;
        currentY++;
        currentP += 2 * delX - 2 * delY;
      }

      console.log(
        `xi = ${currentX}, yi = ${currentY}, ${operatorStr} = ${currentP}`
      );

      this.output.push([currentX, currentY]);
    }

    return this.output;
  }
}

console.log(new Bresenham([2, 3], [8, 6]).run());
console.log(new Bresenham([2, 5], [5, 12]).run());
console.log(new Bresenham([3, 4], [7, 10]).run());
console.log(new Bresenham([2, 3], [8, 7]).run());
